aws-lc-rs/src/unstable/kem.rs

```diff
@@ -0,0 +1,339 @@
+    Kyber768_R3,
+
+    /// NIST Round 3 submission of the Kyber-1024 algorithm.
+    Kyber1024_R3,
+}
+
+impl crate::kem::AlgorithmIdentifier for AlgorithmId {
+    #[inline]
+    fn nid(self) -> i32 {
+        match self {
+            AlgorithmId::Kyber512_R3 => NID_KYBER512_R3,
+            AlgorithmId::Kyber768_R3 => NID_KYBER768_R3,
+            AlgorithmId::Kyber1024_R3 => NID_KYBER1024_R3,
+        }
+    }
+}
+
+impl crate::sealed::Sealed for AlgorithmId {}
+
+/// Retrieve an unstable KEM [`Algorithm`] using the [`AlgorithmId`] specified by `id`.
+/// May return [`Option::None`] if support for the algorithm has been removed from the unstable module.
+#[must_use]
+pub const fn get_algorithm(id: AlgorithmId) -> Option<&'static Algorithm<AlgorithmId>> {
+    match id {
+        AlgorithmId::Kyber512_R3 => Some(&KYBER512_R3),
+        AlgorithmId::Kyber768_R3 => Some(&KYBER768_R3),
+        AlgorithmId::Kyber1024_R3 => Some(&KYBER1024_R3),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        error::KeyRejected,
+        kem::{DecapsulationKey, EncapsulationKey},
+    };
+
+    use super::{get_algorithm, AlgorithmId, KYBER1024_R3, KYBER512_R3, KYBER768_R3};
+
+    #[cfg(private_api)]
+    macro_rules! kem_kat_test {
+        ($file:literal, $alg:expr) => {
+            use crate::{
+                aws_lc::{
+                    pq_custom_randombytes_init_for_testing,
+                    pq_custom_randombytes_use_deterministic_for_testing,
+                },
+                test, test_file,
+            };
+            test::run(test_file!($file), |_section, test_case| {
+                let seed = test_case.consume_bytes("seed");
+                let public_key_bytes = test_case.consume_bytes("pk");
+                let secret_key_bytes = test_case.consume_bytes("sk");
+                let ciphertext_bytes = test_case.consume_bytes("ct");
+                let shared_secret_bytes = test_case.consume_bytes("ss");
+
+                // Set randomness generation in deterministic mode.
+                unsafe {
+                    pq_custom_randombytes_use_deterministic_for_testing();
+                    pq_custom_randombytes_init_for_testing(seed.as_ptr());
+                }
+
+                let priv_key = PrivateKey::generate($alg).unwrap();
+
+                assert_eq!(priv_key.as_ref(), secret_key_bytes);
+
+                let pub_key = priv_key.public_key().unwrap();
+                assert_eq!(pub_key.as_ref(), public_key_bytes);
+
+                let (ciphertext, bob_shared_secret) = pub_key.encapsulate().unwrap();
+                assert_eq!(ciphertext.as_ref(), ciphertext_bytes);
+                assert_eq!(bob_shared_secret.as_ref(), shared_secret_bytes);
+
+                let alice_shared_secret = priv_key.decapsulate(ciphertext).unwrap();
+                assert_eq!(alice_shared_secret.as_ref(), shared_secret_bytes);
+
+                Ok(())
+            });
+        };
+    }
+
+    #[cfg(private_api)]
```

Comment:
Getting these tests to run takes some digging. Should we have a comment here indicating how to setup the build so that these tests execute?
```
AWS_LC_RUST_PRIVATE_INTERNALS=1 cargo test --features unstable,bindgen
```

Reply:
See earlier comment.